### PR TITLE
perf: vite plugin

### PR DESCRIPTION
**Pull request by nonzzz**
# Background

I find the `full-reload` is a bad way :) So i refactor the `hmr` part logic.

## Checklist

- [x] Refactor generator chunk logic.
- [x] Support define file name.
- [x] Perf `hmr` logic.

### CODE REVIEW

File: vite.js [hunk 1]
```diff
@@ -4,91 +4,112 @@ const processCSS = require('./src/process-css');
...
+  return {
+    map,
+    code: output.join('\n')
+  };
+}
+function genCSS(cssModule) {
+  const chunk = Array.from(cssModule.values()).reduce(
+    (acc, cur) => (acc += cur),
+    ''
+  );
```
Comment (johanholmerin) [inline vite.js:56]: Is this not just a more complicated `.join('')`?

File: vite.js [hunk 2]
```diff
@@ -4,91 +4,112 @@ const processCSS = require('./src/process-css');
...
+  if (css) {
+    cssModules.set(id, css);
+    output.unshift(`import '${VIRTUAL_MODULE_NAME}'`);
+    if (server) {
+      const { moduleGraph } = server;
+      const virtualModule = moduleGraph.getModuleById(VIRTUAL_CSS_NAME);
+      if (virtualModule) {
+        const { code } = virtualModule.transformResult;
+        const output = code.split('\n');
+        const finalCoe = output.reduce((acc, cur) => {
```
Comment (johanholmerin) [inline vite.js:37]: This doesn't seem to reduce anything, it's just a map & join

Comment (johanholmerin) [inline vite.js:37]: `finalCoe` typo?

**Review by johanholmerin — CHANGES_REQUESTED** [review]
This seems to be dependent on implementation details of Vite and could easily break?

File: vite.js (hunk 2 above)
Comment (nonzzz) [reply vite.js:37]: It's typo

**Comment by nonzzz** [comment]
@johanholmerin cc.

File: vite.js (hunk 1 above)
Comment (nonzzz) [reply vite.js]: You're right. Code has been simplified :)

**Comment by johanholmerin** [comment]
> This seems to be dependent on implementation details of Vite and could easily break?

**Comment by nonzzz** [comment]
No,It's stable. Because vite's modulegraph can't update virtual module. So i had to implement whit it :)

File: vite.js [hunk 3]
```diff
@@ -4,91 +4,111 @@ const processCSS = require('./src/process-css');
...
+    output.unshift(`import '${VIRTUAL_MODULE_NAME}'`);
+    if (server) {
+      const { moduleGraph } = server;
+      const virtualModule = moduleGraph.getModuleById(VIRTUAL_CSS_NAME);
+      if (virtualModule) {
+        const { code: cssStr } = virtualModule.transformResult;
+        const latest = cssStr
+          .split('\n')
+          .map(str => {
+            if (str.startsWith('const __vite__css')) {
```
Comment (johanholmerin) [inline vite.js:39]: This depends on a line starting with this exact string

Comment (nonzzz) [reply vite.js:39]: Yes. Vite converts css modules to js modules. According https://github.com/vitejs/vite/blob/main/packages/vite/src/node/plugins/css.ts#L398-L411

Comment (johanholmerin) [reply vite.js:39]: Like I said, that's an implementation detail that's not guaranteed to be stable across versions

Comment (nonzzz) [reply vite.js:39]: okay. I'll use another way

**Comment by johanholmerin** [comment]
@nonzzz Is this ready for review?

**Comment by nonzzz** [comment]
> @nonzzz Is this ready for review?

yes.

**Review by johanholmerin — APPROVED** [review]

**Comment by johanholmerin** [comment]
Released in 0.18.0

**Pull request merged**